Sources/MongoSwift/ConnectionString.swift

```diff
@@ -6,52 +6,124 @@ internal class ConnectionString {
             throw extractMongoError(error: error)
         }
         self._uri = uri
 
-        if let rc = options?.readConcern {
-            self.readConcern = rc
+        try self.applyOptions(options)
+    }
+
+    // swiftlint:disable:next cyclomatic_complexity
+    private func applyOptions(_ options: MongoClientOptions?) throws {
+        if let appName = options?.appName {
+            guard mongoc_uri_set_option_as_utf8(self._uri, MONGOC_URI_APPNAME, appName) else {
+                throw MongoError.InvalidArgumentError(message: "Failed to set appName to \(appName)")
+            }
         }
 
-        if let wc = options?.writeConcern {
-            self.writeConcern = wc
+        if let credential = options?.credential {
+            try self.setMongoCredential(credential)
+        }
+
+        // Per SDAM spec: If the ``directConnection`` option is not specified, newly developed drivers MUST behave as
+        // if it was specified with the false value.
+        if let dc = options?.directConnection {
+            guard mongoc_uri_set_option_as_bool(self._uri, MONGOC_URI_DIRECTCONNECTION, dc) else {
+                throw MongoError.InvalidArgumentError(message: "Failed to set directConnection to \(dc)")
+            }
+        } else if !self.hasOption("directConnection") {
+            guard mongoc_uri_set_option_as_bool(self._uri, MONGOC_URI_DIRECTCONNECTION, false) else {
+                throw MongoError.InvalidArgumentError(message: "Failed to set directConnection to false")
+            }
+        }
+
+        if let maxPoolSize = options?.maxPoolSize {
+            // libmongoc casts the value to a uint32_t
+            guard maxPoolSize > 0 && maxPoolSize <= UInt32.max else {
+                throw MongoError.InvalidArgumentError(
+                    message: "Invalid maxPoolSize \(maxPoolSize): must be between 1 and \(UInt32.max)"
+                )
+            }
+            guard mongoc_uri_set_option_as_int32(self._uri, MONGOC_URI_MAXPOOLSIZE, Int32(maxPoolSize)) else {
+                throw MongoError.InvalidArgumentError(message: "Failed to set maxPoolSize to \(maxPoolSize)")
+            }
+        }
+
+        if let rc = options?.readConcern {
+            self.readConcern = rc
         }
 
         if let rp = options?.readPreference {
             self.readPreference = rp
         }
 
-        if let rw = options?.retryWrites {
-            mongoc_uri_set_option_as_bool(self._uri, MONGOC_URI_RETRYWRITES, rw)
+        if let replicaSet = options?.replicaSet {
+            guard mongoc_uri_set_option_as_utf8(self._uri, MONGOC_URI_REPLICASET, replicaSet) else {
+                throw MongoError.InvalidArgumentError(message: "Failed to set replicaSet to \(replicaSet)")
+            }
         }
 
         if let rr = options?.retryReads {
-            mongoc_uri_set_option_as_bool(self._uri, MONGOC_URI_RETRYREADS, rr)
+            guard mongoc_uri_set_option_as_bool(self._uri, MONGOC_URI_RETRYREADS, rr) else {
```

Comment:
turns out these setter methods return bools indicating success, which we have been ignoring 😱